examples/ecs/indexing.rs

```diff
@@ -0,0 +1,205 @@
+            last_this_run: default(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T, F> IndexBacking<T, F> {
+    fn update(&mut self, entity: Entity, value: Option<T>) -> Option<T>
+    where
+        T: Hash + Eq + Clone,
+    {
+        let old = if let Some(ref value) = value {
+            self.reverse.insert(entity, value.clone())
+        } else {
+            self.reverse.remove(&entity)
+        };
+
+        if let Some(ref old) = old {
+            if let Some(set) = self.forward.get_mut(old) {
+                set.remove(&entity);
+
+                if set.is_empty() {
+                    self.forward.remove(old);
+                }
+            }
+        }
+
+        if let Some(value) = value {
+            self.forward.entry(value).or_default().insert(entity);
+        };
+
+        old
+    }
+
+    fn get(&self, value: &T) -> Option<impl Iterator<Item = Entity> + '_>
+    where
+        T: Hash + Eq + Clone,
+    {
+        Some(self.forward.get(value)?.iter().copied())
+    }
+}
+
+#[derive(SystemParam)]
+pub struct Index<'w, 's, T, F = ()>
+where
+    T: Component + Hash + Eq,
+    F: ReadOnlyWorldQuery + 'static,
+{
+    changed: Query<'w, 's, (Entity, Ref<'static, T>), (Changed<T>, F)>,
+    removed: RemovedComponents<'w, 's, T>,
+    index: ResMut<'w, IndexBacking<T, F>>,
+    this_run: SystemChangeTick,
+}
+
+impl<'w, 's, T, F> Index<'w, 's, T, F>
+where
+    T: Component + Hash + Eq + Clone,
+    F: ReadOnlyWorldQuery + 'static,
+{
+    fn update_index_internal(&mut self) {
+        let this_run = self.this_run.this_run();
+
+        // Remove old entires
+        for entity in self.removed.read() {
+            self.index.update(entity, None);
+        }
+
+        // Update new and existing entries
+        for (entity, component) in self.changed.iter() {
+            self.index.update(entity, Some(component.clone()));
+        }
+
+        self.index.last_this_run = Some(this_run);
+    }
+
+    fn update_index(mut index: Index<T>) {
+        index.update_index_internal();
+    }
+
+    fn ensure_updated(&mut self) {
+        let this_run = self.this_run.this_run();
+
+        if self.index.last_this_run != Some(this_run) {
+            self.update_index_internal();
+        }
+    }
+
+    pub fn get(&mut self, value: &T) -> Option<impl Iterator<Item = Entity> + '_> {
+        self.ensure_updated();
+
+        self.index.get(value)
+    }
+
+    pub fn iter(&mut self) -> impl Iterator<Item = Entity> + '_ {
+        self.ensure_updated();
+
+        self.index.reverse.keys().copied()
+    }
+}
+
+pub struct IndexPlugin<T>(PhantomData<T>);
+
+impl<T> Default for IndexPlugin<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T> Plugin for IndexPlugin<T> where T: Component + Hash + Eq + Clone {
+    fn build(&self, app: &mut App) {
+        app.init_resource::<IndexBacking<T>>()
+            .add_systems(Update, Index::<T>::update_index);
+    }
+}
```

Comment:
It seems weird that `Index` has a `F` parameter but not `IndexPlugin`.

Reply:
Yeah that's me running out the door to go out for dinner interrupting a chain of thought! I want the `Index` to support filter queries and custom `Indexer`'s as provided by the user.